src/Agent.Sdk/Util/IOUtil.cs

```diff
@@ -223,6 +227,51 @@ public static void DeleteFile(string path)
             }
         }
 
+        public static async Task DeleteDirectoryWithRetry(string path, CancellationToken cancellationToken)
+        {
+            for (int i = 0; i < MAX_RETRY_DELETION; i++)
+            {
+                try
+                {
+                    DeleteDirectory(path, cancellationToken);
+                    return;
+                }
+                // There is no reason to retry on DirectoryNotFoundException, SecruityException and UnauthorizedAccessException
+                catch (IOException)
+                {
+                    if (!cancellationToken.IsCancellationRequested)
+                    {
+                        // Pause execution briefly to allow the file to become accessible
+                        await Task.Delay(i * 1000, cancellationToken);
+                        continue;
```

Comment:
Bug: this will always 'continue' (assuming no cancelation)

"throw;" will not be hit, even if we hit the retry count limit

retry logic is tricky :) .  This is tricky to unit test but that would be one possible option

Reply:
tested manually (by blocking files and unblocking in a given time, increasing/decreasing retry count) and added unit test